contracts/Plasma.sol

```diff
@@ -0,0 +1,194 @@
+
+        depositCount += 1;
+    }
+
+    // withdrawals
+
+    struct ExitTX {
+        address exitor;
+        uint timestamp;
+    }
+
+    PriorityQueue exitsQueue;
+    mapping(bytes32 => ExitTX) public exits;
+
+    event WithdrawDeposit(address _depositor, uint indexed _depositIndex, bytes32 indexed _uid, uint256 indexed _timestamp, uint _amount);
+    event ExitStarted(uint indexed _priority);
+
+    function withdrawDeposit(uint _depositIndex, address _currency) public {
+        bytes32 uid = depositIndex[_depositIndex];
+        require(uid != 0x0);
+
+        // check if msg.sender is the right one to withdraw
+        require(keccak256(msg.sender, _currency, _depositIndex) == uid);
+
+        ExitTX memory exitTX = ExitTX({
+            exitor: msg.sender,
+            timestamp: block.timestamp + 60
+        });
+
+        addExitToQueue(uid, _depositIndex, exitTX);
+
+        uint amount = depositBalance[uid];
+
+        WithdrawDeposit(msg.sender, _depositIndex, uid, block.timestamp, amount);
+    }
+
+    function addExitToQueue(bytes32 _uid, uint _depositIndex, ExitTX memory exitTX) private {
+        // check if the withdrawal has been already requested
+        require(exits[_uid].timestamp == 0);
+
+        uint priority = exitTX.timestamp << 128 | _depositIndex;
+        exitsQueue.insert(priority);
+        exits[_uid] = exitTX;
+
+        ExitStarted(priority);
+    }
+
+    function getNextExit() public view returns (uint _depositIndex, bytes32 _uid, uint _timestamp) {
+        uint256 priority = exitsQueue.getMin();
+        _depositIndex = uint256(uint128(priority));
+        _uid = depositIndex[_depositIndex];
+        _timestamp = priority >> 128;
+    }
+
+    event FinalizedExit(address _depositor, uint indexed _depositIndex, uint _amount,  bytes32 indexed _uid, uint256 indexed _timestamp);
+
+    function finalizeExits() public returns (uint _num) {
+        uint index;
+        bytes32 uid;
+        uint exitTimestamp;
+
+        uint timestamp = block.timestamp;
+
+        (index, uid, exitTimestamp) = getNextExit();
+        while (exitTimestamp < timestamp) {
+
+            ExitTX memory exitTX = exits[uid];
+
+            uint amount = depositBalance[uid];
+            exitTX.exitor.transfer(amount);
+
+            FinalizedExit(exitTX.exitor, index, amount, uid, timestamp);
+
+            exitsQueue.delMin();
+            _num += 1;
+
+            if (exitsQueue.currentSize() > 0) {
+                (index, uid, exitTimestamp) = getNextExit();
+            } else {
+                return;
+            }
+        }
+    }
+
+    // temp methods
+    function verifyProof(bytes _proof, bytes32 _root, bytes32 _leaf, uint256 _index) public constant returns (bool success) {
+        return MerkleProof.verifyProof(_proof, _root, _leaf, _index);
+    }
+
+    function proveTX(uint _blockIndex, bytes _transactionBytes, bytes _proof) public constant returns (bool success) {
```

Comment:
Unless you're using this on-chain I'd recommend implementing this client-side. Same goes with `proveNoTX`.

Reply:
it's been implemented for testing purposes. Thank you for your point